### ISSUE TITLE
Broken CPU metrics in AWS ECS after updating to CoreOS 1122.2.0

### ISSUE THREAD

**Issue by mlafeldt** [coreos/bugs#1605]
# Issue Report

After updating from CoreOS stable version 1068.9.0 to 1122.2.0, the reported CPU utilization for all AWS ECS services dropped significantly (which in turn broke our EC2 auto scaling).

Our benchmark tests showed that the reported CPU utilization for ECS services differs at least by a factor of 3 between CoreOS versions 1068.9.0 and 1122.2.0. To rule out any effect of the ECS agent, which is the instance reporting CPU utilization to AWS, we ran the benchmark with different ECS agents too. However, changes in the versions show no differences in the metrics:

![coreos-version-benchmark](https://cloud.githubusercontent.com/assets/158074/19270105/9350a46e-8fbe-11e6-805e-1055c9a209eb.png)

This shows the CPU utilization of the affected auto scaling group compared with the ECS cluster. One can see how the ECS metric diverges from the (good) EC2 metric while CoreOS 1122.2.0 was rolled out:

<img width="1033" alt="ecs-cpu-vs-asg-cpu-2" src="https://cloud.githubusercontent.com/assets/158074/19270581/0ee4c978-8fc1-11e6-8e47-889183139e78.png">

The major change between CoreOS versions 1068.9.0 and 1122.2.0 is a new version of the Linux Kernel. Version 1122.2.0 introduced [Kernel version 4.7](https://kernelnewbies.org/Linux_4.7#head-daae031092afaa8b596b233ed38f2bf2b1a5b52d) which adds changes to the cgroup filesystem interface. Those changes add extra information to the `cpuacct` values which are used by all kinds of container tools to monitor resource usage. We suspect the changes to be related to the difference in reported CPU utilization.

Since detecting this problem about a month ago, we've been in contact with AWS support. They claim that "the issue seems specific to CoreOS since it does not happen with the ECS optimized AMI". As we don't want to migrate our cluster fleet to Amazon Linux, we're reaching out to the CoreOS community for help. Maybe we can find a fix together.

Unfortunately, this problem blocks us from updating to any CoreOS version released after 1068.9.0. In fact, I'm sure that CPU metrics are broken for all companies using ECS and CoreOS with newer versions.
### CoreOS Version

Version introducing changes in CPU metrics: CoreOS stable 1122.2.0

Last known good version:

```
NAME=CoreOS
ID=coreos
VERSION=1068.9.0
VERSION_ID=1068.9.0
BUILD_ID=2016-08-05-0455
PRETTY_NAME="CoreOS 1068.9.0 (MoreOS)"
ANSI_COLOR="1;32"
HOME_URL="https://coreos.com/"
BUG_REPORT_URL="https://github.com/coreos/bugs/issues"
```
### Environment

AWS EC2 instances registered with AWS ECS

**Comment by crawford** [comment]
Thank you for the detailed report. I'm going to engage with AWS to see if they are aware of this issue.

**Comment by mlafeldt** [comment]
AFAIK, AWS customer support forwarded this issue to the ECS team. They were able to reproduce it. But that's it.

**Comment by mlafeldt** [comment]
Today, we got word that the ECS team is working on a fix here: https://github.com/aws/amazon-ecs-agent/pull/556

AWS support wrote:

> The team has investigated the problem and will be working on a fix for this. Currently the ECS agent depends on percpu_usage returned by docker stats to get this information, and this is causing the mismatch. The team will be working on using a more reliable method to report this under CoreOS to fix this on the agent side.

**Comment by mlafeldt** [comment]
The fix (https://github.com/aws/amazon-ecs-agent/pull/556) was merged. Now waiting for a new ecs-agent release so we can test it.

**Comment by crawford** [comment]
@mlafeldt thank you for tracking this. I still don't see a release from the ecs-agent team.

**Comment by mlafeldt** [comment]
Sadly, still no new ecs-agent release, even after requesting it in https://github.com/aws/amazon-ecs-agent/issues/576

**Comment by mlafeldt** [comment]
AWS gave us a "pre-released version" of ecs-agent for testing.

We could verify that the fix (aws/amazon-ecs-agent#556) does indeed work. There's no longer a CPU drop when updating to the latest CoreOS stable release:

![screenshot 2016-11-08 15 01 08](https://cloud.githubusercontent.com/assets/158074/20102784/a638ae2c-a5c7-11e6-8364-b380e1bbad8a.png)

Even though there's no official ecs-agent release yet, you can close this issue as far as we're concerned.

**Comment by crawford** [comment]
I'm glad to hear you got it working. Thanks again for tracking the progress on this issue.
